File: stroom-auth-svc/src/integration-test/java/stroom/auth/service/resources/UserResource_update_IT.java
/*
 * Copyright 2017 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stroom.auth.service.resources;

import org.junit.Test;
import stroom.auth.AuthenticationFlowHelper;
import stroom.auth.resources.user.v1.User;
import stroom.auth.service.ApiResponse;
import stroom.auth.service.api.AuthenticationApi;
import stroom.auth.service.api.UserApi;
import stroom.auth.service.api.model.ChangePasswordRequest;
import stroom.auth.service.resources.support.Dropwizard_IT;

import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

public final class UserResource_update_IT extends Dropwizard_IT {
    @Test
    public final void update_user() throws Exception {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());

        ApiResponse<Integer> response = userApi.createUserWithHttpInfo(new stroom.auth.service.api.model.User()
                .email("update_user_" + UUID.randomUUID().toString())
                .password("password"));

        User user = userManager.deserialiseUsers(userApi.getUser(response.getData())).get(0);

        user.setEmail("new_email_" + UUID.randomUUID().toString());

        userApi.updateUser(response.getData(), new stroom.auth.service.api.model.User()
            .id(user.getId())
            .email(user.getEmail()));

        User updatedUser = userManager.deserialiseUsers(userApi.getUser(response.getData())).get(0);

        assertThat(updatedUser.getEmail()).isEqualTo(user.getEmail());
    }

    @Test
    public final void update_self_basic_user() throws Exception {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());

        String userEmailA = "update_user_" + UUID.randomUUID().toString();
        int userEmailAId = userApi.createUserWithHttpInfo(new stroom.auth.service.api.model.User()
                .email(userEmailA)
                .password("password"))
                .getData();

        // If we don't change the password the AuthenticationResource will think this is the first login and
        // try to force us to change the password. This means we won't get an access code and complete the
        // authentication flow. So we'll change the password so the flow completes as normal.
        AuthenticationApi authApi = SwaggerHelper.newAuthApiClient(AuthenticationFlowHelper.authenticateAsAdmin());
        authApi.changePassword(userEmailAId, new ChangePasswordRequest()
                .email(userEmailA)
                .oldPassword("password")
                .newPassword("password"));

        User userA = userManager.deserialiseUsers(userApi.getUser(userEmailAId)).get(0);
        userA.setComments("Updated user");
        UserApi userApiA = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAs(userEmailA, "password"));
        ApiResponse<String> userUpdateResponse = userApiA.updateUserWithHttpInfo(userEmailAId, new stroom.auth.service.api.model.User()
                .id(userA.getId())
                .comments(userA.getComments()));

        assertThat(userUpdateResponse.getStatusCode()).isEqualTo(200);

        User updatedUser = userManager.deserialiseUsers(userApi.getUser(userEmailAId)).get(0);
        assertThat(updatedUser.getComments()).isEqualTo("Updated user");
    }
}

File: stroom-auth-svc/src/integration-test/java/stroom/auth/service/resources/UserResource_update_IT.kt
/*
 * Copyright 2017 Crown Copyright
 *
 * This file is part of Stroom-Stats.
 *
 * Stroom-Stats is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Stroom-Stats is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stroom-Stats.  If not, see <http://www.gnu.org/licenses/>.
 */

package stroom.auth.service.resources

import com.github.kittinunf.fuel.httpPut
import org.assertj.core.api.Assertions
import org.junit.Test
import java.time.Instant

class UserResource_update_IT : UserResource_IT() {

    @Test
    fun update_user() {
        val jwsToken = login()
        val user = User(email = Instant.now().toString(), password = "testPassword")

        // First create a user to update
        var userId = createUser(user, jwsToken)
        user.email = "New email" + Instant.now().toString()
        val serialisedUser = userMapper().toJson(user)

        val url = ROOT_URL + userId
        val (_, updateResponse) = url
                .httpPut()
                .body(serialisedUser)
                .header(jsonContentType("Authorization" to "Bearer $jwsToken"))
                .responseString()
        updateResponse.assertOk()

        var updatedUser = getUser(userId, jwsToken)
        Assertions.assertThat(updatedUser?.email).isEqualTo(user.email)
    }

    //TODO add additional update user tests, e.g. validation issues

}
